File: Recursion/3154. Find Number of Ways to Reach the K-th Stair.cpp
/*
3154. Find Number of Ways to Reach the K-th Stair
*/

// TC:O(2^n) n = 31 包含memo剪枝 SC:O(pos*32*2)
class Solution {
    unordered_map<int, vector<vector<int>>> memo; // pos -> jump狀況 -> 使用的次數
    int res = 0;
public:
    int waysToReachStair(int k) {
        
        
        
        DFS(1, 0, 0, k);
        
        return res;
        
    }
    int DFS(int pos, int jump, bool rev, int k){
        if(jump > 30) return 0;
        if(memo.find(pos) != memo.end() && memo[pos][jump][rev] != -1) return memo[pos][jump][rev];
        int count = (pos == k);
        
        if(pos != 0 && rev == 0){
            count += DFS(pos-1, jump, 1, k);
        }
        if(pos + (1<<jump) <= k+1)
            count += DFS(pos+(1<<jump), jump+1, 0, k);
        
        memo[pos].resize(32, vector<int>(2,-1));
        memo[pos][jump][rev] = count;
        res = max(res, count);
        
        return count;
        
    }
};

/*
從1 開始跳
1. 往回走一格，但不能連續用or 在stair 0 上用
2. 走到i+ 2^j格, j = j + 1, initial j = 0;
問最後alice 有多少種方法走到k格上 2^30次方可以是10^9
1 + ? = 10^9 30跳的時候 肯定會爆炸超過 2^29 = 5e8, 2^30 = 1e9 up
只有當跳到比k大1以上的格子就是結束 -> BFS or DP+memo
BFS -> TLE
但目前問題是pos位置太大了1e9 所以不可能每個位置都存-> Hash Map 再來就是條件上的不同 jump狀況以及是否能往回走
原來resize 可以直接做二維
*/

File: Dynamic_Programming/3154. Find Number of Ways to Reach the K-th Stair.py
# DP TC:O(lgk * lgk * 2) SC:O(lg k * lg k * 2)
class Solution:
    def waysToReachStair(self, k: int) -> int:
        '''
        init: pos = 1, jump = 0
        每次可以做的操作是 往下走一階 但不能連續用 or  < 0 
        或者可以選擇走 i + 2 ^ jump，然後j + 1
        問說有多少種方式可以走到 k 的位置
        因為每個狀態如果都包含進來就可以讓當下狀態的結果有唯一性 所以可以考慮dfs + memo 的dp
        dfs (i, j, choose) 這樣設計
        每次都可能會走 2^n 次方 所以可以假設我可以用lgk的次數走到該位置，也可能往前走，但畢竟限制往回走次數不能連續
        這個間接的指明不可能倒退走，所以就算往前都倒退走 那也就是 lg(k + lgk) 大概這麼多步
        '''
        @cache 
        def dfs(i:int, j:int, choose:bool) -> int:
            if i > k + 1:
                return 0
            ans = 1 if i == k else 0
            if not choose and i - 1 >= 0:
                ans += dfs(i - 1, j, True)
            ans += dfs(i + (1 << j), j + 1, False)
            return ans
        return dfs(1, 0, False)

# 組合數學 TC: O(1) SC:O(1)
class Solution:
    def waysToReachStair(self, k: int) -> int:
        '''
        init: pos = 1, jump = 0
        每次可以做的操作是 往下走一階 但不能連續用 or  < 0 
        或者可以選擇走 i + 2 ^ jump，然後j + 1
        問說有多少種方式可以走到 k 的位置
        因為每個狀態如果都包含進來就可以讓當下狀態的結果有唯一性 所以可以考慮dfs + memo 的dp
        dfs (i, j, choose) 這樣設計
        每次都可能會走 2^n 次方 所以可以假設我可以用lgk的次數走到該位置，也可能往前走，但畢竟限制往回走次數不能連續
        這個間接的指明不可能倒退走，所以就算往前都倒退走 那也就是 lg(k + lgk) 大概這麼多步
        '''
        ans = 0
        start = (max(0,k-1)).bit_length()
        for j in range(start, 30):
            m = (1 << j) - k
            if m > j + 1:
                break
            if m <= j + 1:
                ans += comb(j + 1, m) #這玩意可以預處理 用遞推式 C n取m = Cn-1 取 m -1 + c n-1 取 m  選or不選去思考得出
        return ans
